Sources/ApolloSQLite/SQLiteNormalizedCache.swift

```diff
@@ -64,17 +57,8 @@ public final class SQLiteNormalizedCache {
     return updatedComponents.joined(separator: ".")
   }
 
-  private func createTableIfNeeded() throws {
-    try self.db.run(self.records.create(ifNotExists: true) { table in
-      table.column(id, primaryKey: .autoincrement)
-      table.column(key, unique: true)
-      table.column(record)
-    })
-    try self.db.run(self.records.createIndex(key, unique: true, ifNotExists: true))
-  }
-
   private func mergeRecords(records: RecordSet) throws -> Set<CacheKey> {
```

Comment:
Hello, since I'm a recent Apollo user, I dare to chime in!

This `mergeRecords` method performs several database operations.

If it were wrapped in a transaction, SQLite would be faster. I do not know if it is a concern.

If it were wrapped in a transaction, you'd get never get a half-modified database on disk, should an error happen at some point during the method execution. I also do not know if it is a concern.

If it were wrapped in a transaction, eventual concurrent reads would not see a half-modified database. Maybe a concern, or not.

Reply:
This is the pinnacle of open source collaboration. <3 
A maintainer of a popular open source repo being a user of another repo _using_ their repo. And making suggestions to improve the usage of their repo using their knowledge of both of them. :) Haha thanks @groue this is a great comment.

Reply:
I guess technically we aren't using GRDB _yet_ .... but we are getting there. >_<

Reply:
That's why this pull request is a very good idea :-) I tried to ask questions in a way that is independent from the preferred adapter. I surely don't want to imply anything else or to influence your decision. It's just that defining an adapter protocol is a good opportunity to wonder about how Apollo interacts with its cache, and document the expected behavior, as you said above.

Reply:
Do you think that the `SQLiteDatabase` protocol should have some concept of a transaction built-in? I imagine that most any SQLLite implementation would likely have transactions and being able to utilize them via the inverted dependency protocol would be valuable.

Reply:
I don't know for sure. I mentioned `mergeRecords` because it looks like a method that could enjoy a transaction (for speed, or in order to enforce integrity guarantees). Transactions, when needed, are usually defined by the database client (here, Apollo), because only the database client knows how individual database operations should be grouped. Maybe Apollo does not need any transaction! But if it does, then I would indeed ask the adapter to provide transaction boundaries.

Yes, all "serious" SQLite wrappers make it able to perform transactions. APIs fall generally into three groups:

1. Block-based transactions

    ```swift
    // Form 1
    wrapper.transaction {
         wrapper.doStuff()
    }

    // Form 2
    wrapper.transaction { dbAccess in
         dbAccess.doStuff()
    }

    // Form 3
    wrapper.perform {
         wrapper.doStuff()
         wrapper.save()
    }
    ```

    SQLite.swift and Realm are form 1. FMDB and GRDB are form 2. Core Data is form 3.

2. "Free-form" transactions

    ```swift
    wrapper.beginTransaction()
    wrapper.doStuff()
    wrapper.commit()
    ```

    Most SQLite implementations that allow raw SQL are here. Some kind of concurrency control must be added.

3. Transaction objects

    ```swift
    let transaction = wrapper.transaction()
    transaction.doStuff()
    transaction.commit()
    ```

    Core Data can be used like this, by detaching an NSManagedObjectContext. GRDB as well, by creating a new DatabaseQueue. OK, this is really not frequent.

Some wrappers distinguish *read* from *write* transaction. This is an opportunity to improve performance by allowing multiple readers with a single writer.

I think you can accept all forms above with an adapter that looks like:

```swift
adapter.write { transaction in
    transaction.doStuff()
}
```

Something should serialize writes (either the adapter contract, or Apollo).

If Apollo needs to read data, with a concern about cache integrity, and a need for multiple readers and single writer for adapters that support it, then a read transaction is useful as well:

```swift
let data = adapter.read { transaction in
    transaction.loadStuff()
}
```

---

This only applies if you think Apollo needs transactions.

Reply:
@groue @AnthonyMDev 🎶 It's the ciiiiiiiircle of liiiiiiiife... 🎶

I think i probably want to leave transaction support out for this PR since i'm just trying to do a 1:1 "what are we already doing" conversion before adding any further complexity. 

However, you're absolutely correct that this would be way better, particularly from a standpoint of write-ahead log support (which is better for things with extensions etc). I'll take a look at that in my next PR.